### DISCUSSION TITLE
[Vote] Runout Sensor

### DISCUSSION THREAD

**Discussion by Jyers** [Jyers/Marlin#481] [Ideas]
Unfortunately due to marlins limitations, I have to have the pin state for the runout sensor set on compile. This means there has to be one pin state for all the precompiled versions. I've gotten a few discussions of people who have it inverted and don't know how to compile. That being said, I feel I should check what the most common config is and make that the default.

### High: 🚀 Low: 👀

**Comment by dwidou** [comment]
Low for me. Maybe would it be possible to have a check box in "control" with by exemple "High by defaut, check if sensor low" ?

**Reply by dwidou** [reply]
ANd thanks for your awesome Ender 3 v2 firmware !!! ;-)

**Reply by Jyers** [reply]
Unfortunately as I said in the post, the pin state has to be set on compile. There is not way for me to modify it after the fact via an on screen setting

**Comment by rakosi2** [comment]
I am using this mod which would make the FIL_RUNOUT_STATE HIGH. I can always just swap the wires around If needed. There isn't much documentation on Runout Sensors so if you create a wiki page about it and how to invert, it should be fine.
https://notenoughtech.com/3dprint/filament-runout-sensor-for-ender3-v2/

**Comment by G4GNX** [comment]
I'm using the BigTreeTech sensor which sends pulses. ISTR they're positive going.

**Comment by duncio** [comment]
> Unfortunately due to marlins limitations, I have to have the pin state for the runout sensor set on compile. This means there has to be one pin state for all the precompiled versions. I've gotten a few discussions of people who have it inverted and don't know how to compile. That being said, I feel I should check what the most common config is and make that the default.
> 
> ### What is your `FIL_RUNOUT_STATE`?
> ### High: 🚀 Low: 👀

@Jyers Seems it might be possible -> https://www.facebook.com/groups/ender3v2/permalink/807185963227758/

**Comment by Cal-LF** [comment]
Another option is to physically swap wire connections on the switch type filament runout sensor.
As shown above by Miguel Risco-Castillo, his filament runout sensor menu looks like a solution.

**Reply by Jyers** [reply]
To achieve this he most likely had to modify some base marlin files. I honestly don't know what he's doing at this point. Originally he looked like he was using mostly base marlin and adding a few tings that could get merged, but now he's just porting features from this firmware into the old dwin code, while having seemingly no intent on creating a PR.

**Reply by G4GNX** [reply]
ISTM that if this is not a genuine Marlin distro or official branch like yours, it's best to stay away.
I'm more than happy with what you do and if I want something special, I'll compile my own, with yours as a base and keep it to myself. :-)

**Comment by Dylotic** [comment]
Mine is Low

**Comment by Cal-LF** [comment]
I have been compiling the 3X3 manual mesh so either is ok for me. Those using the precompiled still can rewire the switch to make it work.

**Comment by hunkin1969** [comment]
Actually for me its working opposite, so i cant use it. If i press down the switch (with filament), i have contact and printer tells me Filament off, buts filled. So this means i need high?

**Reply by Cal-LF** [reply]
Can the wires be reversed on the switch?

**Reply by hunkin1969** [reply]
I tried, but then its like a reset switch. The whole system starts new. Thats my switch 
![Endstop](https://user-images.githubusercontent.com/82881357/115397611-94724980-a1e6-11eb-9672-fdffc229f5b3.JPG)

**Reply by Dylotic** [reply]
> I tried, but then its like a reset switch. The whole system starts new. Thats my switch 
> 
> ![Endstop](https://user-images.githubusercontent.com/82881357/115397611-94724980-a1e6-11eb-9672-fdffc229f5b3.JPG)
> 
> 

I've had the same problem. I needed to turn around the two wires on the outside.

I found the solution in the last comment on this forum:

https://forum.creality.com/topic/767/ender-3-v2-card-v4-2-2-filament-runout-wiring/6?lang=nl

In your schematic the red wire (+5v) should be on top connection on the board in the picture.

I think now you are shorting the board by activating the switch. This results in a bad reset.

I pushed out the pins and switched them around. Now everything works perfectly

To invert the action of the switch you need to compile the firmware with LOW/HIGH

**Reply by hunkin1969** [reply]
Yes, i know this post and this problem. I allready changed the pins before.

**Reply by Dylotic** [reply]
> Yes, i know this post and this problem. I allready changed the pins before. 
> 
> 

Then leave it original and adjust the firmware FIL_RUNOUT_STATE to HIGH

**Reply by hunkin1969** [reply]
Yes, i shoul. If i knew how to do... Many thanks for trying to help. But at the moment im not able to do such firmware mods. I will wait for a solution.

**Comment by smooreace** [comment]
The solution is tuning the firmware to fit your needs, or modify your needs
to fit the firmware.

**Comment by eecue** [comment]
I switched the pins to work with what you have currently, so don't change it, or if you do let me know so I can tweak my firmware build!

**Reply by hunkin1969** [reply]
Due i am at the moment not able to change anything on the firmware, i would highly be thankful if you would do so. I downloaded all things i basically need for firmware compiling, but it seems, im not the programmer guy. I wanted to learn, but i failed in the first steps. Maybe the instructions are not good, or i am too stupid for this. LOL. I think second one.....

**Comment by Woodcat64** [comment]
I think High would be better. As it is now, if the sensor is disconnected printer thinks filament is still there.
